packages/framework/presence/src/container-definitions/containerExtensions.ts

```diff
@@ -0,0 +1,162 @@
+	/**
+	 * Notifies the extension of a new use context.
+	 *
+	 * @param context - Context new reference to extension is acquired within
+	 */
+	onNewContext(...context: TContext): void;
+
+	/**
+	 * Callback for signal sent by this extension.
+	 *
+	 * @param address - Address of the signal
+	 * @param signal - Signal content and metadata
+	 * @param local - True if signal was sent by this client
+	 */
+	processSignal?(address: string, signal: IExtensionMessage, local: boolean): void;
+}
+
+/**
+ * Defines the runtime interface an extension may access.
+ * In most cases this is a subset of {@link @fluidframework/container-runtime-definitions#IContainerRuntime}.
+ *
+ * @sealed
+ * @internal
+ */
+export interface IExtensionRuntime {
+	/**
+	 * {@inheritdoc @fluidframework/container-runtime-definitions#IContainerRuntime.clientId}
+	 */
+	get clientId(): ClientConnectionId | undefined;
+
+	/**
+	 * Submits a signal to be sent to other clients.
+	 * @param address - Custom address for the signal.
+	 * @param type - Custom type of the signal.
+	 * @param content - Custom content of the signal. Should be a JSON serializable object or primitive via {@link https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify|JSON.stringify}.
+	 * @param targetClientId - When specified, the signal is only sent to the provided client id.
+	 *
+	 * Upon reciept of signal, {@link IContainerExtension.processSignal} will called with the same
+	 * address, type, and content (less any non-{@link https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify|JSON.stringify}-able data).
+	 */
+	submitAddressedSignal<T>(
+		address: string,
+		type: string,
+		content: JsonSerializable<T>,
+		targetClientId?: ClientConnectionId,
+	): void;
+}
+
+/**
+ * Factory method to create an extension instance.
+ *
+ * Any such method provided to {@link ContainerExtensionStore.acquireExtension}
+ * must use the same value for a given {@link ContainerExtensionId} so that an
+ * `instanceof` check may be performed at runtime.
+ *
+ * @typeParam T - Type of extension to create
+ * @typeParam TContext - Array of optional custom context
+ *
+ * @param runtime - Runtime for extension to work against
+ * @param context - Custom context for extension.
+ * @returns Record providing:
+ * `extension` instance (type `T`) that is provided to caller of
+ * {@link ContainerExtensionStore.acquireExtension} and
+ * `interface` store/runtime uses to interact with extension.
+ *
+ * @internal
+ */
+export type ContainerExtensionFactory<T, TContext extends unknown[]> = new (
+	runtime: IExtensionRuntime,
+	...context: TContext
+) => { readonly extension: T; readonly interface: IContainerExtension<TContext> };
+
+/**
+ * Unique identifier for extension
+ *
+ * @remarks
+ * A string known to all clients working with a certain ContainerExtension and unique
+ * among ContainerExtensions. Recommend using specifying concatenation of: type of
```

Comment:
nit: "Recommend using specifying.." is kind of confusing, but I may just be misunderstanding